### PR TITLE
feat: allow fs embedding with `--embed`

### LINKED ISSUE #130: Support for embedding static files at build-time
Since [Caddy v2.6.0](https://github.com/caddyserver/caddy/releases/tag/v2.6.0), we have virtual file systems support, which also means `file_server` can serve files embedded in the binary via https://pkg.go.dev/embed.

We have a proof-of-concept module at https://github.com/mholt/caddy-embed which can do this, but I don't think this is ideal UX.

Instead, I think we should support a `xcaddy --embed path/to/embed/dir/` flag which would copy the files into the temp build directory, embed them into the binary via code generation, and transparently add a module to the binary that implements `caddy.fs.embedded`.

I wrote a bit about this some time ago in https://github.com/mholt/caddy-embed/issues/1, but I'm opening the issue here since it's a better place to track it.

### PR DESCRIPTION
**Pull request by mohammed90**
This PR adds the ability to embed a folder into the compiled Caddy by copying the entire tree passed via `--embed` (or `Builder#EmbedDir` struct field) that's specified into a directory named `files` under the temporary build directory and synthesize a Caddy module in the same dir (a copy of github.com/mholt/caddy-embed). This function relies on [Go's `embed` package](https://pkg.go.dev/embed) for embedding capability, thus it abides by the rules listed in its documentation.

~~I couldn't figure out to how to do multiple folders to allow the file server to do serve different roots, hence~~ it's marked Experimental 🤷🏼 

Resolves #130 
Resolves mholt/caddy-embed#1

### CODE REVIEW

**Comment by mholt** [comment]
Oh neat!! This is cool, I'll give it a look soon :)

**Comment by francislavoie** [comment]
I'm not sure how embedding works, but can we have like a `dir:` kind of prefix for multiple dirs?

**Comment by mohammed90** [comment]
> I'm not sure how embedding works, but can we have like a `dir:` kind of prefix for multiple dirs?

Now that I stepped back and thought about it, the current implementation can support it if the user provides the dirs already nested. In other words, if they embed `sites` which contains:
- `sites/example.com`
- `sites/example.net`

They can do

```
example.com , example.net
 file_server {
    fs embedded
    root {host}
}
```

If the user specifies multiple `--embed`, we build the nesting similarly instead of the user building it beforehand.

One thing I'd like to address is to somehow expose the embedded FS tree so users can inspect their fat binaries. That can be in a follow-up.

**Comment by francislavoie** [comment]
Can we allow the user to pass multiple `--embed` and if so nest them ourselves with those directory names?

File: internal/utils/io.go
Comment (mholt) [inline internal/utils/io.go:1]: Can we possibly just put this in the xcaddy package and leave it unexported? (Can still be in its own file.)

**Review by mholt — COMMENTED** [review]
Thanks, this could be a neat/useful enhancement!

**Comment by mohammed90** [comment]
> Can we allow the user to pass multiple `--embed` and if so nest them ourselves with those directory names? Maybe by default we should ask the user to provide a directory name. We could use a syntax like `--embed embed-name:./path/to/dir` or something.

Done. You can now specify multiple `--embed name1:path/to/dir` or `--embed path/to/dir`, and we'll take charge of the nesting. If no name is specified, the embedded dir content is placed at the root of the tree.

File: internal/utils/io.go
Comment (mohammed90) [reply internal/utils/io.go:1]: Done.

**Comment by francislavoie** [comment]
Nice! :clap:

File: environment.go [hunk 1]
```diff
@@ -86,11 +86,41 @@ func (b Builder) newEnvironment(ctx context.Context) (*environment, error) {
...
+				return nil, err
+			}
+			_, err = os.Stat(d.Dir)
+			if err != nil {
+				return nil, fmt.Errorf("embed directory does not exist: %s", d.Dir)
+			}
+			log.Printf("[INFO] Embedding directory: %s", d.Dir)
+
+			var buf bytes.Buffer
+			tpl, err := template.New("embed").Parse(embeddedModuleTemplate)
```
Comment (mohammed90) [inline environment.go:107]: Don't merge yet. I need to fix the `err` shadowing here, which prevents the deferred cleanup because it's not the same `err.

**Comment by mohammed90** [comment]
> So I assume this works?

Yep, yep, and yep. 

> Also I think we're missing support for `fs` for `try_files` in Caddy?

Hm... perhaps, I haven't inspected the extenet of that particular change in Caddy. We can create an issue to track it.

**Comment by mholt** [comment]
This is neat!! Is it ready to go?

**Comment by mohammed90** [comment]
> This is neat!! Is it ready to go?

Yep!

**Comment by mholt** [comment]
Thanks for working on this :) Let's give it a try

**Comment by francislavoie** [comment]
> `try_files` supports an embedded FS.

https://caddyserver.com/docs/caddyfile/directives/try_files doesn't though I think (easy fix though).

And I think we should use a var or whatever to bubble it down to other modules similarly to `root`. Usability will be not great from a config standpoint until we do that.

**Comment by korpa** [comment]
I really like this feature. But sadly for example directories like `_app` (used by svelte) are not included. Would it be possible to support files and dirs starting with underscores as well?

See https://github.com/golang/go/issues/43854

**Comment by korpa** [comment]
Furthermore I wonder if I could use embedded files with cli command only instead of using `Caddyfile`: 


Some sort of `caddy files_server --embed --listen :8080`

**Comment by mohammed90** [comment]
> Would it be possible to support files and dirs starting with underscores as well?

It may be possible to include a condition/flag to have the `all:` prefix for the pattern, but I'm not sure what the best UX is for that. I don't want to introduce another flag that depends on this flag. I also don't want to add embedded flags to this pattern ` <[alias]:path/to/dir>...`. You can always embed the directory with a name that does not start with an underscore then rewrite the request path to have the underscore. How does this sound?



> Furthermore I wonder if I could use embedded files with cli command only instead of using `Caddyfile`:
> 
> Some sort of `caddy files_server --embed --listen :8080`

We cannot add a flag to a command that's defined elsewhere (i.e. the file-server module in this case). I personally believe one must bite the bullet at some point and switch to config files, i.e. Caddyfile, rather than craft long lines of commands and flags. If we are to add a flag, it'd be `--fs` rather than `--embed`, but again this should be discussed in the caddyserver/caddy repo, not here.

**Pull request merged**